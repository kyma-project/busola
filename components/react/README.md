# Kyma React Components

## Overview

This application is a ReactJS component library for the Kyma, based on the [component-library-starter](https://github.com/alanbsmith/component-library-starter).

You can see the components and their properties on the [Busola](https://kyma-project.github.io/console) page.

## Usage

To import the library into an application, run the following command:

```bash
npm install --save @kyma-project/react-components
```

## Development

### Build

After the build, the webpack creates the `lib` folder.

Run once:

```bash
npm run build
```

Then run the `watch` script:

```bash
npm run build:watch
```

### Test

**Jest Snapshots** manages all of the testing. **Jest Snapshots** places the tests in the `__tests__` folders, on the component level.

To run the tests once:

```bash
npm test
```

To run the watch script:

```bash
npm run test:watch
```

To view the coverage report:

```bash
npm run test:coverage:report
```
