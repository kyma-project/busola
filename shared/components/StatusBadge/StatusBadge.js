import React from 'react';
import { useTranslation } from 'react-i18next';
import { ObjectStatus } from 'fundamental-react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import './StatusBadge.scss';
import { TooltipBadge } from '../TooltipBadge/TooltipBadge';

const resolveType = status => {
  if (typeof status !== 'string') {
    console.warn(
      `'autoResolveType' prop requires 'children' prop to be a string.`,
    );
    return undefined;
  }

  switch (status.toUpperCase()) {
    case 'INITIAL':
    case 'PENDING':
      return 'informative';

    case 'READY':
    case 'RUNNING':
    case 'SUCCESS':
    case 'SUCCEEDED':
    case 'OK':
      return 'positive';

    case 'UNKNOWN':
    case 'WARNING':
      return 'critical';

    case 'FAILED':
    case 'ERROR':
    case 'FAILURE':
    case 'INVALID':
      return 'negative';

    default:
      return undefined;
  }
};

const translate = (i18n, arrayOfVariableNames, fallbackValue) => {
  if (!i18n) return fallbackValue;
  const { t } = useTranslation(null, { i18n });
  return t(arrayOfVariableNames, { defaultValue: fallbackValue });
};

const prepareTranslationPath = (resourceKind, value, type) => {
  return `${resourceKind.toString().toLowerCase()}.${type}.${value
    .toString()
    .toLowerCase()
    .replace(/\s/g, '-')}`;
};

const TYPE_FALLBACK = new Map([
  ['success', 'positive'],
  ['warning', 'critical'],
  ['error', 'negative'],
  ['info', 'informative'],
]);

export const StatusBadge = ({
  additionalContent,
  tooltipContent, // deprecated
  type,
  resourceKind = 'common',
  children: value = '',
  autoResolveType = false,
  tooltipProps = {},
  noTooltip = false,
  className,
  i18n,
}) => {
  if (autoResolveType) type = resolveType(value);
  else
    for (const key of TYPE_FALLBACK.keys()) {
      if (type === key) type = TYPE_FALLBACK.get(key);
    }
  const i18nFullVariableName = prepareTranslationPath(
    resourceKind,
    value,
    'statuses',
  );
  const commonStatusVariableName = prepareTranslationPath(
    'common',
    value,
    'statuses',
  );
  const tooltipVariableName = prepareTranslationPath(
    resourceKind,
    value,
    'tooltips',
  );
  const commonTooltipVariableName = prepareTranslationPath(
    'common',
    value,
    'tooltips',
  );
  const fallbackValue = value.toString();
  const classes = classNames(
    'status-badge',
    {
      'has-tooltip': !noTooltip,
    },
    className,
  );

  const badgeContent = translate(
    i18n,
    [i18nFullVariableName, commonStatusVariableName],
    fallbackValue,
  );

  let content = translate(
    i18n,
    [tooltipVariableName, commonTooltipVariableName, i18nFullVariableName],
    fallbackValue,
  );

  if (additionalContent) {
    content = `${content}: ${additionalContent}`;
  }

  // tooltipContent is DEPRECATED. Use the TooltipBadge component if a Badge with a simple Tooltip is needed.
  if (tooltipContent) {
    return (
      <TooltipBadge
        tooltipContent={tooltipContent}
        type={type}
        tooltipProps={tooltipProps}
        className={classes}
      >
        {badgeContent}
      </TooltipBadge>
    );
  } else if (noTooltip) {
    return (
      <ObjectStatus
        ariaLabel="Status"
        role="status"
        inverted
        status={type}
        className={classes}
        data-testid="no-tooltip"
      >
        {badgeContent}
      </ObjectStatus>
    );
  } else {
    return (
      <TooltipBadge
        tooltipContent={content}
        type={type}
        tooltipProps={tooltipProps}
        className={classes}
      >
        {badgeContent}
      </TooltipBadge>
    );
  }
};

StatusBadge.propTypes = {
  additionalContent: PropTypes.node,
  tooltipContent: PropTypes.node,
  type: PropTypes.oneOf([
    'positive',
    'negative',
    'critical',
    'informative',
    'success',
    'error',
    'warning',
    'info',
  ]),
  autoResolveType: PropTypes.bool,
  noTooltip: PropTypes.bool,
  resourceKind: PropTypes.string,
  tooltipProps: PropTypes.object,
  i18n: PropTypes.object,
  className: PropTypes.string,
};
