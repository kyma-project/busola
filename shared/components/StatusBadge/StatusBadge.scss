.status-badge {
  text-transform: uppercase;
}

.has-tooltip {
  cursor: help;
}
