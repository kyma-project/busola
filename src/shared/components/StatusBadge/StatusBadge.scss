.status-badge {
  text-transform: uppercase;
  word-break: keep-all;
}
