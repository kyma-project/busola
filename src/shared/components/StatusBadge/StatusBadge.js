import React from 'react';
import { useTranslation } from 'react-i18next';
import { ObjectStatus } from '@ui5/webcomponents-react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import './StatusBadge.scss';
import { TooltipBadge } from 'shared/components/TooltipBadge/TooltipBadge';

const resolveType = status => {
  if (typeof status !== 'string') {
    console.warn(
      `'autoResolveType' prop requires 'children' prop to be a string.`,
    );
    return undefined;
  }

  switch (status.toUpperCase()) {
    case 'INITIAL':
    case 'PENDING':
    case 'AVAILABLE':
    case 'RELEASED':
      return 'Information';

    case 'READY':
    case 'BOUND':
    case 'RUNNING':
    case 'SUCCESS':
    case 'SUCCEEDED':
    case 'OK':
      return 'Success';

    case 'UNKNOWN':
    case 'WARNING':
      return 'Warning';

    case 'FAILED':
    case 'ERROR':
    case 'FAILURE':
    case 'INVALID':
      return 'Error';

    default:
      return 'None';
  }
};

const translate = (i18n, t, arrayOfVariableNames, fallbackValue) => {
  return i18n.exists(arrayOfVariableNames)
    ? t(arrayOfVariableNames)
    : fallbackValue;
};

const prepareTranslationPath = (resourceKind, value, type) => {
  return `${resourceKind.toString().toLowerCase()}.${type}.${value
    .toString()
    .toLowerCase()
    .replace(/\s/g, '-')}`;
};

const TYPE_FALLBACK = new Map([
  ['success', 'Success'],
  ['warning', 'Warning'],
  ['error', 'Error'],
  ['info', 'Information'],
]);

export const StatusBadge = ({
  additionalContent,
  tooltipContent, // deprecated
  type,
  resourceKind = 'common',
  children: value = '',
  autoResolveType = false,
  tooltipProps = {},
  noTooltip = false,
  className,
}) => {
  const { t, i18n } = useTranslation();
  if (autoResolveType) type = resolveType(value);
  else
    for (const key of TYPE_FALLBACK.keys()) {
      if (type === key) type = TYPE_FALLBACK.get(key);
    }
  const i18nFullVariableName = prepareTranslationPath(
    resourceKind,
    value,
    'statuses',
  );
  const commonStatusVariableName = prepareTranslationPath(
    'common',
    value,
    'statuses',
  );
  const tooltipVariableName = prepareTranslationPath(
    resourceKind,
    value,
    'tooltips',
  );
  const commonTooltipVariableName = prepareTranslationPath(
    'common',
    value,
    'tooltips',
  );
  const fallbackValue = value.toString();
  const badgeContent = translate(
    i18n,
    t,
    [i18nFullVariableName, commonStatusVariableName],
    fallbackValue,
  );
  let content = translate(
    i18n,
    t,
    [tooltipVariableName, commonTooltipVariableName, i18nFullVariableName],
    fallbackValue,
  );

  if (!additionalContent && content === badgeContent) noTooltip = true;

  const classes = classNames(
    'status-badge',
    {
      'has-tooltip': !noTooltip,
    },
    className,
  );

  if (additionalContent) {
    if (badgeContent === content) content = additionalContent;
    else {
      // Remove the dot at the end of the sentence if we add a colon afterwards
      if (content?.endsWith('.')) content = content?.slice(0, -1);
      content = `${content}: ${additionalContent}`;
    }
  }

  // tooltipContent is DEPRECATED. Use the TooltipBadge component if a Badge with a simple Tooltip is needed.
  if (tooltipContent) {
    return (
      <TooltipBadge
        tooltipContent={tooltipContent}
        type={type}
        tooltipProps={tooltipProps}
        className={classes}
      >
        {badgeContent}
      </TooltipBadge>
    );
  } else if (noTooltip) {
    return (
      <ObjectStatus
        aria-label="Status"
        role="status"
        inverted
        state={type}
        className={classes}
        data-testid="no-tooltip"
      >
        {badgeContent}
      </ObjectStatus>
    );
  } else {
    return (
      <TooltipBadge
        tooltipContent={content}
        type={type}
        tooltipProps={tooltipProps}
        className={classes}
      >
        {badgeContent}
      </TooltipBadge>
    );
  }
};

StatusBadge.propTypes = {
  additionalContent: PropTypes.node,
  tooltipContent: PropTypes.node,
  type: PropTypes.oneOf(['information', 'success', 'error', 'warning', 'none']),
  autoResolveType: PropTypes.bool,
  noTooltip: PropTypes.bool,
  resourceKind: PropTypes.string,
  tooltipProps: PropTypes.object,
  className: PropTypes.string,
};
