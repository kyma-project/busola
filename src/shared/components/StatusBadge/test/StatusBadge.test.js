import React from 'react';
import { StatusBadge } from 'shared/components/StatusBadge/StatusBadge';
import { act, render, waitFor } from '@testing-library/react';
import { ThemeProvider } from '@ui5/webcomponents-react';

describe('StatusBadge', () => {
  it('renders status text with proper role', async () => {
    const { queryByRole } = render(
      <ThemeProvider>
        <StatusBadge>INITIAL</StatusBadge>
      </ThemeProvider>,
    );

    await waitFor(async () => {
      await act(async () => {
        const status = queryByRole('status');
        expect(status).toBeInTheDocument();
        expect(status).toHaveTextContent('common.statuses.initial');
      });
    });
  });

  it('displays warning when autoResolveType is set and "children" is a node', async () => {
    console.warn = jest.fn();

    render(
      <ThemeProvider>
        <StatusBadge autoResolveType>
          <small>Status</small>
        </StatusBadge>
      </ThemeProvider>,
    );
    await waitFor(async () => {
      await act(async () => {
        expect(console.warn.mock.calls[0]).toMatchSnapshot();
      });
    });
  });

  it('renders status text with DEFAULT_STATUSES_PATH', async () => {
    const DEFAULT_STATUSES_PATH = 'common.statuses.initial';
    // 'common.statuses.initial,common.statuses.initial,fallback';
    const { queryByRole } = render(
      <ThemeProvider>
        <StatusBadge>Initial</StatusBadge>
      </ThemeProvider>,
    );
    await waitFor(async () => {
      await act(async () => {
        const status = queryByRole('status');
        expect(status).toBeInTheDocument();
        expect(status).toHaveTextContent(DEFAULT_STATUSES_PATH);
      });
    });
  });

  it('renders status text with RESOURCE_STATUSES_PATH', async () => {
    const RESOURCE_KIND = 'resource';
    const RESOURCE_STATUSES_PATH = 'resource.statuses.initial';
    const { queryByRole } = render(
      <ThemeProvider>
        <StatusBadge resourceKind={RESOURCE_KIND}>Initial</StatusBadge>
      </ThemeProvider>,
    );
    await waitFor(async () => {
      await act(async () => {
        const status = queryByRole('status');
        expect(status).toBeInTheDocument();
        expect(status).toHaveTextContent(RESOURCE_STATUSES_PATH);
      });
    });
  });

  it('renders status text without tooltip', async () => {
    const { getByTestId } = render(
      <ThemeProvider>
        <StatusBadge noTooltip>Initial</StatusBadge>
      </ThemeProvider>,
    );
    await waitFor(async () => {
      await act(async () => {
        const status = getByTestId('no-tooltip');
        expect(status).toBeInTheDocument();
      });
    });
  });

  it('renders status text with tooltip', async () => {
    const { getByTestId } = render(
      <ThemeProvider>
        <StatusBadge>Initial</StatusBadge>
      </ThemeProvider>,
    );
    await waitFor(async () => {
      await act(async () => {
        const status = getByTestId('has-tooltip');
        expect(status).toBeInTheDocument();
      });
    });
  });
});
