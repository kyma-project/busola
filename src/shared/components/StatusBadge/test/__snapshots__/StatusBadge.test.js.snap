// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StatusBadge > displays warning when autoResolveType is set and "children" is a node 1`] = `
[
  "'autoResolveType' prop requires 'children' prop to be a string.",
]
`;
